node/overseer/src/lib.rs

```diff
@@ -919,15 +953,15 @@ where
 	async fn block_imported(&mut self, block: BlockInfo) -> SubsystemResult<()> {
 		let mut update = ActiveLeavesUpdate::default();
 
-		if let Some(parent) = block.number.checked_sub(1).and_then(|number| self.active_leaves.take(&(block.parent_hash, number))) {
-			update.deactivated.push(parent.0);
+		if let Some(_number) = self.active_leaves.remove(&block.parent_hash) {
```

Comment:
Probably a good idea to remove the vector, if present, from `activation_external_listeners` here as well, as a first-line anti-clutter defense. Might be worth factoring a `self.on_head_deactivated` method similar to `self.on_head_activated`.

Reply:
yeah, I initially made response_channel for `activation_external_listeners` return a `bool` instead of `()` signalling if the block was deactivated, but there was two problems with this
1. in `block_finalized`, it's not possible to call a method inside of `retain` due to borrowing rules, this is more an annoyance than the problem
2. if the block was deactivated before the subscription, we will never respond, so for the caller it won't make much difference because it has to handle timeouts anyway.